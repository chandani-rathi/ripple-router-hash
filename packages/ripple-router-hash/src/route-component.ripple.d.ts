import type { Component } from "ripple";

export type DynamicDataProps = {
    $component: Function, 
    $data: {[key:string]: string}, 
    $children: Component
}

export declare function DynamicDataComponent({ $component, $data, $children }): Component

export type RouteProp = {
    id: string,
    path: string,
    children: RouteProp[]
    import: () => any
    props: {[key:string]: string}
}

export declare function RouteComponent({ $component }: { $component: RouteProp}): Component