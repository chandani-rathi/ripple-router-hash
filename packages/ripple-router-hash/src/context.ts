import { createContext } from "ripple";

export const RouteContext = createContext(null);