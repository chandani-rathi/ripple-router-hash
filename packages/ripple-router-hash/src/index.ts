export * from "@/router-app.ripple";
export * from "@/route.ripple";
export * from "@/route-component.ripple";

export * from "@/components/link.ripple";
export * from "@/components/not-found.ripple";
export * from "@/components/loading.ripple";
export * from "@/components/error.ripple";


export * from "@/state.ripple";
export * from "@/lib/utils";
export * from "@/lib/hash-router";