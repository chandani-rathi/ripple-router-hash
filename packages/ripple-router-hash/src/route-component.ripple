import { effect, TrackedArray , track } from 'ripple';
import { getHashPath, arrayToTree, findRoute } from "@/lib/utils";
import { NotFound } from "@/components/not-found.ripple";
import { Loading } from "@/components/loading.ripple";
import { Error } from "@/components/error.ripple";
import { RouteContext } from "@/state.ripple";


function getComponent(module) {
	const mainComponent =  module.Layout || module.Page || module.default || null;
	const loadingComponent = module.Loading;
	const errorComponent = module.Error;
	const loader = module.loader;
	const metadata = module.metadata;
	return {
		mainComponent, loadingComponent, errorComponent, loader, metadata
	}
}

function updateMetadata(metadata, context){
	if(metadata) {
		document.title = metadata.title
	}
	else if(context?.documentTitle){
		document.title = context.documentTitle
	}
}

export function isPromiseLike(value) {
  return value && typeof value.then === "function";
}

export component DynamicDataComponent({ $component, data, children}: { $component, data?, children?}){
	let $dynamicComponent = (__anchor, props, __block) => {
		console.log("loading dynamic data", @data)
		$component(__anchor, {...props, ...(@data || {})}, __block)
	};

	<$dynamicComponent>
		if(children){
			<children />
		}
	</$dynamicComponent>
}

export component RouteComponent({ id, route, routeProps, routeChildren  }){
	const loading = track(true);
	let $loadingComponent = null;

	let $error = false;
	let $errorComponent = null;

	let routeComponent = track(null);
	let $childComponentLoaded = null;

	let context = RouteContext.get(RouteContext);

	const routeData = track({...(routeProps || {})});

	effect(() => {
		if(!id == "not-found") {
			@loading = false;
			return
		};

		try {
			route()
				.then(getComponent)
				.then(module => {
					updateMetadata(module.metadata, context);
					if(module.loadingComponent) {
						console.log("page has loading component", id);
						$loadingComponent = module.loadingComponent;
					}
					if(module.errorComponent) {
						console.log("page has error component", id);
						$errorComponent = module.errorComponent;
					}
					return module
				})
				.then((module) => {
					return new Promise((resolve, reject) => {
						if(module.loader){
							const loaderData = typeof(module.loader) === "function" ? module.loader({...(routeProps || {})}): {};
							if(isPromiseLike(loaderData)) {
								loaderData.then(r => resolve({...module, loaderData: r})).catch(reject)
							}
							else{
								resolve({...module, loaderData})
							}
						}
						else {
							resolve({...module, loaderData: {}})
						}
					})
				})
				.then(({mainComponent, loaderData}) => {
						if(!mainComponent && id != "not-found") {
							console.warn("warning: node named Page or default export found in page ", $component.props)
						}
						@loading = false;
						@routeData = {...(routeProps || {}), loaderData};
						@routeComponent = mainComponent;
						if(id.startsWith("layout:")) {
							$childComponentLoaded = setTimeout(() => {
								console.warn("warning: make sure to user children in layout component ", $component.id)
							}, 1000)
						}
					})
					.catch(e => {
						console.error("Error occurred while loading route. Error Code: 1", routeProps);
						console.error(e);
						$error = true;
					});
		}
		catch(e) {
			console.error("Error occurred while loading route. Error Code: 2", routeProps);
			console.error(e)
		}
	})

	try {
		if(@loading) {
			if($error) {
				if($errorComponent) {
					<DynamicDataComponent $component={$errorComponent} data={routeData}/>
				}
				else {
					<Error />
				}
			}
			else if($loadingComponent) {
				<DynamicDataComponent $component={$loadingComponent} data={routeData}/>
			}
			else {
				<Loading />
			}
		}
		else if(id == "not-found") {
			<NotFound />
		}
		else if(@routeComponent){
			<DynamicDataComponent $component={@routeComponent} data={routeData}>
				component children() {
					clearTimeout($childComponentLoaded)
					// if(routeChildren.length == 1) {
					// 	<RouteComponent $component={routeChildren[0]} />
					// }
					// if(routeChildren.length == 2) {
					// 	<RouteComponent $component={routeChildren[1]} />
					// }
					// if(routeChildren.length == 3) {
					// 	<RouteComponent $component={routeChildren[2]} />
					// }
					for(let child of routeChildren) {
						<RouteComponent id={child.id} route={child.route} routeProps={child.props} routeChildren={child.children} />
					}
				}
			</DynamicDataComponent>
		}
	
	}
	catch(e) {
		console.error("Error occurred while loading route. Error Code: 3", routeProps);
		console.error(e)
		if($errorComponent) {
			try {
				<DynamicDataComponent $component={$errorComponent} data={routeData}/>
			}
			catch(e) { 
				console.error("Error occurred while loading route. Error Code: 4", routeProps);
				console.error(e);
				<Error />
			}
		}
		else {
			<Error />
		}
	}
}