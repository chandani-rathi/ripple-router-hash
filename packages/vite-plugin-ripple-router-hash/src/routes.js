import path from "node:path";
import fg from "fast-glob";

export function generateRoutesTree(routes) {
  const routeTree = {
    id: "root",
    path: "",
    import: null,
    children: []
  };

  if (!routes || routes.length == 0) {
    return routeTree;
  }

  for (let route of routes) {
    const path = route.path;
    const segments = path.split("/").filter(Boolean);

    let parentRoute = routeTree;
    let currentId = routeTree.path;
    for (let part of segments) {
      const foundRoute = parentRoute.children.find(f => f.id == `${currentId}/${part}`);
      if (!foundRoute) {
        parentRoute.children.push({
          id: path,
          path: `${part}`,
          import: route.import,
          children: []
        })
      }
      else {
        parentRoute = foundRoute;
        currentId = `${currentId}/${part}`
      }
    }
  }

  return routeTree;
}

function fileToRoutePath(file, fileBasedRoutes) {
  let route = "/";
  if (fileBasedRoutes) {
    route = "/" + file.replace(/\.ripple$/, "");
    route = route.replace(/\/index$/, "");
    route = route.replace(/\[([^\]]+)\]/g, ":$1");
    return route === "/index" ? "/" : route;
  }
  else {
    route = "/" +  file.replace(/\/page.ripple$/, "");
    if(route === "/page.ripple") {
      route = "/"
    }
    route = route.replace(/\[([^\]]+)\]/g, ":$1");
    return route;
  }
  
  return (route === "/index" || route === "/page.ripple") ? "/" : route;
}

function getLayoutChain(file, layouts) {
  // Normalize to forward slashes for cross-platform consistency
  const normalizedFile = file.replace(/\\/g, "/");
  const dir = path.posix.dirname(normalizedFile);
  const dirParts = dir === "." ? [] : dir.split("/").filter(Boolean);

  const chain = [];
  let currentPath = "";

  // ✅ Check root layout first
  if (layouts.has("")) {
    chain.push({
      id: "",
      import: layouts.get("")
    });
  }

  // ✅ Walk subdirectories
  for (const part of dirParts) {
    currentPath = currentPath ? `${currentPath}/${part}` : part;
    if (layouts.has(currentPath)) {
      chain.push({
        id: currentPath,
        import: layouts.get(currentPath)
      });
    }
  }
  return chain;
}

export function generateRoutes({ pagesDir, fileBasedRoutes }) {
  const pageFilter = fileBasedRoutes ? "**/*.ripple" : "**/page.ripple";
  const ignorePageFilter = fileBasedRoutes ? ["**/_*.ripple"] : [];
  const pageFiles = fg.sync(pageFilter, { cwd: pagesDir, ignore: ignorePageFilter });

  const layoutFilter = fileBasedRoutes ? "**/_layout.ripple" : "**/layout.ripple";
  const layoutFiles = fg.sync(layoutFilter, { cwd: pagesDir });

  const layouts = new Map();
  layoutFiles.forEach((file) => {
    const dir = path.posix.dirname(file.replace(/\\/g, "/"));
    const key = dir === "." ? "" : dir; // ✅ root layout key is ""
    layouts.set(key, `() => import("/${pagesDir}/${file}")`);
  });

  return pageFiles.map((file) => {
    const routePath = fileToRoutePath(file, fileBasedRoutes);
    const layoutChain = getLayoutChain(file, layouts);
    return `{
        path: "${routePath}",
        component: () => import("/${pagesDir}/${file}"),
        layouts: [${layoutChain.map(l => `{id: "${l.id}", import: ${l.import} }`).join(", ")}],
      }`;
  });
}