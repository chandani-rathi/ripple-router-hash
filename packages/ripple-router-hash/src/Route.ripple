import { effect, track, RippleArray } from 'ripple';
import { arrayToTree } from "@/lib/utils";
import { NotFound } from "@/components/not-found.ripple";
import { Loading } from "@/components/loading.ripple";
import { RouteComponent } from "./route-component.ripple"

export component Route({ route }: {
	route?: { 
		layouts?: any[],
		component: Function,
		params?: {[key: string]: string},
		path: string
	}}) {
	const isRouteNotFound = @track(false);
	const isRouteLoading = @track(true);
	const newTree = @track(null);
	const tree = new RippleArray();

	effect(() => {
		if(!@newTree) return;

		console.log("loading new route component", @newTree);
		const loadedRoutes = JSON.parse(sessionStorage.getItem("loadedRoutes") || "[]")
		console.log(loadedRoutes);

		const visitNode = (node, previousNode, currentTree) => {
			console.log("node", node);
			console.log("previousNode", previousNode);
			if(previousNode && previousNode.id != node.id) {
				console.log("removing previous node and adding new node")
				const newNode = { ... node}
				newNode.children = new RippleArray();
				currentTree.splice(0, 1, newNode);
			}
			else if(previousNode && previousNode.id == node.id) {
				console.log("both previous node and new node is same")
				//tree.splice(i, 1, node);
			}
			else {
				console.log("adding new node");
				const newNode = { ... node}
				newNode.children = new RippleArray();
				currentTree.push(newNode);
			}

			if(node.children.length > 0) {
				const childNode = node.children[0]
				const childPreviousNode = (previousNode && previousNode.children[0] ) ?? null;
				visitNode(childNode, childPreviousNode, currentTree[0].children);
			}
			
		}

		console.log(JSON.stringify(tree));
		let depth = 0;
		const getDepth = (node) => {
			if(!node || !node[0]) 
				return

			const selectedNode = node[0];
			depth = depth + 1;
			getDepth(selectedNode.children)
		}
		getDepth(loadedRoutes)
		const previousRoutesDepth = depth;

		depth = 0;
		getDepth(@newTree)
		const newRoutesDepth = depth;


		depth = 0;
		const removeNodeFromDepth = (node, startingDepth) => {
			if(!node || !node[0]) 
				return
			const selectedNode = node[0];
			depth = depth + 1;
			if(depth > startingDepth){
				node.splice(0, 1)
				return
			}
			getDepth(selectedNode.children)
		}

		console.log("depth", previousRoutesDepth, newRoutesDepth)
		
		if(previousRoutesDepth <= newRoutesDepth) {
			const node = @newTree[0]
			const previousNode = loadedRoutes ? loadedRoutes[0] : null;
			visitNode(node, previousNode, tree);
		}
		else {
			removeNodeFromDepth(tree, newRoutesDepth);
			console.log("modified tree", JSON.stringify(tree, null ,2 ));
			const node = @newTree[0]
			const previousNode = loadedRoutes ? loadedRoutes[0] : null;
			visitNode(node, previousNode, tree);
		}
		
		console.log(JSON.stringify(tree, null ,2 ))
		sessionStorage.setItem("loadedRoutes", JSON.stringify(@newTree));
	})

	effect(() => {
		if(!route || typeof route.path == "undefined" ) {
			@isRouteLoading = false;
			@isRouteNotFound = true;
			return
		};
		const comps = []
		const { layouts, component, params, path} =  route;
		const props = { params, path }
		if(layouts?.length > 0) {
			layouts.forEach(l => comps.push({ id: `layout:${l.id}`, "route": l.import, props: props}))
		}
		if(route.path && component) {
			comps.push({id: `page:${route.path}`, "route": component, props: props});
		}
		else {
			comps.push({id: `not-found`, "route": null, props: props});
		}
		@isRouteLoading = false;

		const newTreeRoute = arrayToTree(comps);
		@newTree = newTreeRoute;
		// console.log(newTree)
		// @tree = newTree;
	})

	if (@isRouteLoading) {
		<Loading />
	}
	
	if (@isRouteNotFound) {
		<NotFound />
	} 
	
	for(let child of tree) {
		<RouteComponent id={child.id} route={child.route} routeProps={child.props} routeChildren={child.children} />
	}
}