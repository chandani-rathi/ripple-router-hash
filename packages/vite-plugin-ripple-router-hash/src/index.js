import { generateRoutes, generateRoutesTree } from "./routes.js";

export default function rippleRoutesPlugin(options = { pagesDir : "src/pages", fileBasedRoutes: false}) {
  const pagesDir = options.pagesDir || "src/pages";
  const fileBasedRoutes = options.fileBasedRoutes;
  const virtualModuleId = "virtual:ripple-routes";
  const resolvedVirtualId = "\0" + virtualModuleId;

  let routeCode = "";

  function rebuild() {
    console.log("fileBasedRoutes", fileBasedRoutes)
    const routes = generateRoutes({ pagesDir, fileBasedRoutes });
    const routesTree = {}//generateRoutesTree(routes);
    routeCode = `export default [${routes.join(",\n")}]; \n\n export const routesTree = ${JSON.stringify(routesTree, null, 4)}`;
  }

  return {
    name: "vite-plugin-ripple-routes",

    buildStart() {
      rebuild();
    },

    handleHotUpdate(ctx) {
      if (ctx.file.includes("page.ripple") || ctx.file.includes("layout.ripple")) {
        rebuild();
        ctx.server.ws.send({ type: "full-reload", path: "*" });
      }
    },

    resolveId(id) {
      if (id === virtualModuleId) return resolvedVirtualId;
    },

    load(id) {
      if (id === resolvedVirtualId) return routeCode;
    }
  };
}


