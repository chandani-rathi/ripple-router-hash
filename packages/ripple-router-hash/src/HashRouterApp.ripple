import { effect, untrack, mount, createContext  } from 'ripple';
import { getHashPath, findRoute } from "@/lib/utils";
import { NotFound } from "@/components/NotFound.ripple";
import { Loading } from "@/components/Loading.ripple";
import routes from 'virtual:ripple-routes';
import { Route } from "./Route.ripple";
import { RouteContext } from "./context";

export function createHashRouterApp({ target}) {
	console.log("App Routes: ", routes);
	mount(HashRouterApp, { target});
}


export component HashRouterApp() {
	let $currentPath = untrack(() => getHashPath());
	let $loading = true;
	let $newRoute = null;

	effect(() => {
		$loading = false;
		$newRoute = findRoute($currentPath, routes);
	});

	untrack(() => {
		RouteContext.set({ documentTitle: document.title });
	})

	effect(() => {
		
		const onPopState = () => {
			$currentPath = getHashPath();
		};
		window.addEventListener('hashchange', onPopState);

		return () => window.removeEventListener('hashchange', onPopState);
	});
	
	if($loading){
		<Loading />
	}
	else if (!$newRoute) {
		<NotFound />
	} 
	else {
		<Route $route={$newRoute} />
	}
}
