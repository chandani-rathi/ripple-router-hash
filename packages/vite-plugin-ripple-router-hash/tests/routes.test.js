import { expect, it, describe} from 'vitest'
import { generateRoutesTree } from '../src/routes';

describe('routes', () => {

  it('generate routes tree on empty array', () => {
    const routes = []
    const tree = generateRoutesTree(routes);
    console.log(tree)
    expect(1).toBe(1)
  });

  it('generate routes tree on routes', () => {
    const routes = [
    {
        "path": "/about",
        "layouts": [
            {
                "id": ""
            }
        ]
    },
    {
        "path": "/",
        "layouts": [
            {
                "id": ""
            }
        ]
    },
    {
        "path": "/blog",
        "layouts": [
            {
                "id": ""
            }
        ]
    },
    {
        "path": "/blog/:slug",
        "layouts": [
            {
                "id": ""
            }
        ]
    },
    {
        "path": "/dashboard",
        "layouts": [
            {
                "id": ""
            },
            {
                "id": "dashboard"
            }
        ]
    },
    {
        "path": "/dashboard/settings",
        "layouts": [
            {
                "id": ""
            },
            {
                "id": "dashboard"
            }
        ]
    },
    {
        "path": "/dashboard/users",
        "layouts": [
            {
                "id": ""
            },
            {
                "id": "dashboard"
            }
        ]
    },
    {
        "path": "/dashboard/users/:id",
        "layouts": [
            {
                "id": ""
            },
            {
                "id": "dashboard"
            }
        ]
    }
]
    const tree = generateRoutesTree(routes);
    console.log(JSON.stringify(tree, null, 2))
    expect(1).toBe(1)
  })
});
