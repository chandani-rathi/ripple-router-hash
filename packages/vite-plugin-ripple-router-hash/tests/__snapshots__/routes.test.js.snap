// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`routes > generate routes tree on routes 1`] = `
{
  "children": [
    {
      "children": [],
      "id": "/about",
      "import": undefined,
      "path": "about",
    },
    {
      "children": [
        {
          "children": [],
          "id": "/blog/:slug",
          "import": undefined,
          "path": ":slug",
        },
      ],
      "id": "/blog",
      "import": undefined,
      "path": "blog",
    },
    {
      "children": [
        {
          "children": [],
          "id": "/dashboard/settings",
          "import": undefined,
          "path": "settings",
        },
        {
          "children": [
            {
              "children": [],
              "id": "/dashboard/users/:id",
              "import": undefined,
              "path": ":id",
            },
          ],
          "id": "/dashboard/users",
          "import": undefined,
          "path": "users",
        },
      ],
      "id": "/dashboard",
      "import": undefined,
      "path": "dashboard",
    },
  ],
  "id": "root",
  "import": null,
  "path": "",
}
`;
